vpc/tool/container2/setup_container_linux.go

```diff
@@ -254,52 +249,169 @@ func configureLink(ctx context.Context, nsHandle *netlink.Handle, link netlink.L
-		return false, nil
-	} else if err != nil {
-		return false, err
+
+	err = netns.Set(netns.NsHandle(netnsfd))
+	if err != nil {
+		return fmt.Errorf("Could not switch net ns: %w", err)
 	}
-	return true, nil
-}
 
-func waitForAddressUp(ctx context.Context, nsHandle *netlink.Handle, link netlink.Link) error {
-	ctx, cancel := context.WithTimeout(ctx, networkSetupWait)
-	defer cancel()
-	pollTimer := time.NewTicker(100 * time.Millisecond)
-	defer pollTimer.Stop()
+	networkInterface, err := net.InterfaceByName("eth0")
+	if err != nil {
+		return fmt.Errorf("Could not get network interface eth0: %w", err)
+	}
+
+	conn, _, err := ndp.Dial(networkInterface, ndp.Addr(linkLocalAddr.IP.String()))
+	if err != nil {
+		return fmt.Errorf("Unable to dial NDP conn: %w", err)
+	}
+	defer conn.Close()
 
+	var filter ipv6.ICMPFilter
+	filter.SetAll(true)
+	filter.Accept(ipv6.ICMPTypeRouterAdvertisement)
+	err = conn.SetICMPFilter(&filter)
+	if err != nil {
+		return fmt.Errorf("Unable to setup ICMP filter %v: %w", filter, err)
+	}
+
+	m := &ndp.RouterSolicitation{
+		Options: []ndp.Option{
+			&ndp.LinkLayerAddress{
+				Direction: ndp.Source,
+				Addr:      networkInterface.HardwareAddr,
+			},
+		},
+	}
+
+	/* Poll for the router */
+	var routerAddr net.IP
 	for {
-		select {
-		case <-ctx.Done():
-			return errAddressSetupTimeout
-		case <-pollTimer.C:
-			if found, err := isIPv6Ready(nsHandle, link); err != nil {
-				return err
-			} else if found {
-				return nil
-			}
+		err = ctx.Err()
+		if err != nil {
+			return fmt.Errorf("Context complete before NDP finished: %w", err)
 		}
+
+		err = conn.WriteTo(m, &ipv6.ControlMessage{
+			HopLimit: 1,
+			Src:      linkLocalAddr.IP,
+			Dst:      net.IPv6linklocalallrouters,
+			IfIndex:  networkInterface.Index,
+		}, net.IPv6linklocalallrouters)
+		if err != nil {
+			return fmt.Errorf("Could not write router solicitication message: %w", err)
+		}
+
+		// TODO: Maybe don't hardcode this?
+		deadline := time.Now().Add(time.Second)
+		if ctxDeadline, ok := ctx.Deadline(); ok && ctxDeadline.Before(deadline) {
+			deadline = ctxDeadline
+		}
+		err = conn.SetReadDeadline(deadline)
+		if err != nil {
+			return fmt.Errorf("Unable to set read deadline to %v: %w", deadline, err)
+		}
+
+		srcMsg, srcCM, srcAddr, err := conn.ReadFrom()
+		if err == nil {
+			// We're guaranteed only to get router advertisements because of the filter we setup
+			logger.G(ctx).WithFields(map[string]interface{}{
+				"msg":     fmt.Sprintf("%+v", srcMsg),
+				"srcCM":   fmt.Sprintf("%+v", srcCM),
+				"srcAddr": srcAddr,
+			}).Info("Received router advertisement")
+			routerAddr = srcAddr
+			break
+		}
+		if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+			logger.G(ctx).Debug("I/O read timeout")
+			continue
+		}
+		return fmt.Errorf("Could not read router soliticiation response / router advertisement: %w", err)
 	}
+
+	for _, route := range allocation.Routes {
+		if route.Family != vpcapi.AssignIPResponseV3_Route_IPv6 {
```

Comment:
This confuses me.

Are we going to get `allocation.Routes` that are ipv6, but we won't want them? Is this like a feature flag?

Reply:
hrm? this is equivalent to putting the for loop in a route.family == IPv6 {
...add route..
}

Reply:
Oh ok. I couldn't tell what `AssignIPResponseV3_Route_IPv6` is. Did it just get added to the api definitions?

Reply:
yep.

Reply:
it's how enums work in protobuf 🤷